node/setup.go

```diff
@@ -571,20 +572,30 @@ func LoadStateFromDBOrGenesisDocProvider(
 		return sm.State{}, nil, err
 	}
 
-	if err := csGenDoc.GenesisDoc.ValidateAndComplete(); err != nil {
+	if err = csGenDoc.GenesisDoc.ValidateAndComplete(); err != nil {
 		return sm.State{}, nil, fmt.Errorf("error in genesis doc: %w", err)
 	}
 
 	if len(genDocHash) == 0 {
 		// Save the genDoc hash in the store if it doesn't already exist for future verification
-		if err := stateDB.SetSync(genesisDocHashKey, csGenDoc.Sha256Checksum); err != nil {
+		if err = stateDB.SetSync(genesisDocHashKey, csGenDoc.Sha256Checksum); err != nil {
 			return sm.State{}, nil, fmt.Errorf("failed to save genesis doc hash to db: %w", err)
 		}
 	} else {
 		if !bytes.Equal(genDocHash, csGenDoc.Sha256Checksum) {
 			return sm.State{}, nil, fmt.Errorf("genesis doc hash in db does not match loaded genesis doc")
 		}
 	}
+
+	// Validate that existing or recently saved genesis file hash matches optional --genesis_hash passed by operator
+	recentGenDoc, err := stateDB.Get(genesisDocHashKey)
+	if err != nil {
+		return sm.State{}, nil, fmt.Errorf("error retrieving genesis doc hash: %w", err)
+	}
+	if len(storeConfig.GenesisHash) != 0 && !bytes.Equal(recentGenDoc, storeConfig.GenesisHash) {
```

Comment:
I wonder whether, in case they are not the same ,we should delete the hash from the DB to make sure there is not an incorrect hash in the database left in case of a mismatch.

Reply:
Yes, that would be good. Yesterday testing something else on this branch like other genesis I had to reset the DB because the hash mismatch but not sure it is a possible scenario

Reply:
I believe this was suggested here https://github.com/cometbft/cometbft/pull/1324/files/d2021d1aeac819d7079a0820775a37f3617e15e4#r1322667963 but that has been marked as resolved without commiting the suggested change so I'm assuming that you decided that this is not necessary